server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/configuration/ConfigurationAggregate.java

```diff
@@ -0,0 +1,124 @@
+            Preconditions.checkNotNull(maybeConfiguration);
+            this.maybeConfiguration = maybeConfiguration;
+        }
+
+        boolean set(CassandraMailQueueViewConfiguration configuration) {
+            validateConfiguration(configuration);
+            boolean isSame = maybeConfiguration
+                .map(currentConfiguration -> currentConfiguration.equals(configuration))
+                .orElse(false);
+
+            if (!isSame) {
+                maybeConfiguration = Optional.of(configuration);
+                return true;
+            } else {
+                return false;
+            }
+        }
+
+        private void validateConfiguration(CassandraMailQueueViewConfiguration configurationUpdate) {
+            Preconditions.checkNotNull(configurationUpdate);
+
+            maybeConfiguration.ifPresent(currentConfiguration -> {
+                Preconditions.checkArgument(configurationUpdate.getBucketCount() >= currentConfiguration.getBucketCount(),
+                    "can not set 'bucketCount'(" + configurationUpdate.getBucketCount() + ") to be less than the current one: "
+                        + currentConfiguration.getBucketCount());
+
+                long updateSliceWindowInSecond = configurationUpdate.getSliceWindow().getSeconds();
+                long currentSliceWindowInSecond = currentConfiguration.getSliceWindow().getSeconds();
+                Preconditions.checkArgument(
+                    updateSliceWindowInSecond <= currentSliceWindowInSecond
+                        && currentSliceWindowInSecond % updateSliceWindowInSecond == 0,
+                    "update 'sliceWindow'(" + configurationUpdate.getSliceWindow() + ") have to be less than and divisible by the current one: "
+                        + currentConfiguration.getSliceWindow());
+            });
+        }
+    }
+
+    static ConfigurationAggregate load(History history) {
+        return new ConfigurationAggregate(history);
+    }
+
+    private static final String CONFIGURATION_AGGREGATE_KEY = "CassandraMailQueueViewConfiguration";
+    static final AggregateId CONFIGURATION_AGGREGATE_ID = () -> CONFIGURATION_AGGREGATE_KEY;
+    private static final List<? extends Event> EMPTY_EVENTS = ImmutableList.of();
+
+    private final History history;
+    private final State state;
+
+    ConfigurationAggregate(History history) {
+        this.history = history;
+        this.state = State.initial();
+
+        history.getEvents().forEach(this::apply);
+    }
+
+    List<? extends Event> applyConfiguration(CassandraMailQueueViewConfiguration configuration) {
+        ConfigurationEdited newEvent = new ConfigurationEdited(
+            history.getNextEventId(),
+            configuration);
+
+        return apply(newEvent);
+    }
+
+    Optional<CassandraMailQueueViewConfiguration> getCurrentConfiguration() {
+        return state.maybeConfiguration;
+    }
+
+    private List<? extends Event> apply(Event event) {
```

Comment:
We apply command, not events!

Reply:
Wrong.
apply is here to update the State, which is a projection. A projection deals with Event. The aggregate deals with Command. So this apply method is (almost) ok.